lightning/src/onion_message/messenger.rs

```diff
@@ -176,26 +221,57 @@ pub enum SendError {
 ///
 /// [`IgnoringMessageHandler`]: crate::ln::peer_handler::IgnoringMessageHandler
 /// [`CustomMessage`]: Self::CustomMessage
-pub trait CustomOnionMessageHandler {
+pub trait CustomOnionMessageHandler: ResponseErrorHandler {
 	/// The message known to the handler. To support multiple message types, you may want to make this
 	/// an enum with a variant for each supported message.
 	type CustomMessage: CustomOnionMessageContents;
-	/// Called with the custom message that was received.
-	fn handle_custom_message(&self, msg: Self::CustomMessage);
+
+	/// Called with the custom message that was received, returning a response to send, if any.
+	fn handle_custom_message(&self, msg: Self::CustomMessage) -> Option<Self::CustomMessage>;
+
 	/// Read a custom message of type `message_type` from `buffer`, returning `Ok(None)` if the
 	/// message type is unknown.
 	fn read_custom_message<R: io::Read>(&self, message_type: u64, buffer: &mut R) -> Result<Option<Self::CustomMessage>, msgs::DecodeError>;
 }
 
-impl<ES: Deref, NS: Deref, L: Deref, CMH: Deref> OnionMessenger<ES, NS, L, CMH>
-	where ES::Target: EntropySource,
-		  NS::Target: NodeSigner,
-		  L::Target: Logger,
-		  CMH::Target: CustomOnionMessageHandler,
+/// Handler for errors occurring when responding to an onion message.
+///
+/// Each message handler may choose to reply to an onion message with a response. This handler is
+/// used when an error occurs when responding.
+pub trait ResponseErrorHandler {
```

Comment:
So `handle_custom_message` returns `Option<Self::CustomMessage, OnionMessagePath>` and then the messenger is just responsible for `send_onion_message`? Agree that would simplify things - eg if you can't find a route to the intro node, your handler can just give up.

IIUC, I think some form of `ResponseErrorHandler` would still be needed to notify if `send_onion_message` fails so that your handler knows what to do next?

Reply:
> > FYI, another reason I did it this way is so logging was within OnionMessenger.
> 
> I'm not convinced this should be a goal - it costs nothing to add a logger to the downstream implementation (eg the ChannelManager will already have one), and if it leads to avoiding a lot of strange API it seems worth it?

The problems isn't with having a logger, but rather with implementations needing to write the code that logs in five places in `respond_with_onion_message` across however many handler traits there happen to be going forward.

> More broadly, the current PR doesn't actually implement `handle_response_error` anywhere, and its not super clear to me what we can do with the response error but log anyway? Like, if we receive an invoice_request and fail to send the invoice back, it may be worth logging, but we can't actually do anything - the payment may still complete with a second invoice_request received.

Yeah, it was just going to log in the `ChannelManager` implementation, anyhow. I'm inclined to drop `ResponseErrorHandler` for now and keep the interface as is. We can always rethink it later if there is some concrete action the user could take.

Reply:
Removed `ResponseErrorHandler`.